torch/csrc/api/include/torch/data/samplers/stream.h

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <torch/data/samplers/base.h>
+#include <torch/data/samplers/custom_index.h>
+#include <torch/tensor.h>
+
+#include <cstddef>
+
+namespace torch {
+namespace serialize {
+class InputArchive;
+class OutputArchive;
+} // namespace serialize
+} // namespace torch
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+/// A wrapper around a batch size value, which implements the `CustomIndex`
+/// interface.
+struct BatchSize : public CustomIndex {
+  explicit BatchSize(size_t size);
+  size_t size() const noexcept override;
+  operator size_t() const noexcept;
+  size_t size_;
+};
```

Comment:
I don't think this API is very well suited towards streaming datasets. Basically, I would think of such a dataset as an (potentially infinite) list that's read-once. Once you pop an element, it's gone forever. However, if there's a deterministic process that generates that stream, I'd imagine that no matter how I get to the `n`-th index, the sample that's returned will always be the same. If I understand correctly, this implementation simply ignores the indices completely, and takes whatever first `batch_size` elements at each step are available?

That would be fine if a user implemented this, but it's not fine in this case I think. It would be much better to represent a `CustomIndex` for this dataset as `base` offset + `batch_size` of elements that should be taken from the stream (and the stream asserts that at next iter `base` will point to `base + batch_size` from this one).

Reply:
> deterministic process

is a pretty hard constraint here, no? This `StreamDataset` should work with something like `/etc/urandom` -- no determinism at all, infinite, and all we do is ask for `batch_size` elements. The `index` kept in the `StreamDataset` is purely to keep track of progress to know how many elements were sampled. I will rename it to `progress` in fact. Do you see what I mean? What do you think?

Reply:
Saying that the nice properties in the deterministic case don't hold in the non-deterministic case (and are equivalent to this), and concluding that we're going to go with something that doesn't ever work is weird tbh.

Reply:
Ok. What you're really describing is that what we call an index (which is some kind of a unique identifier for a dataset element; a key if you will), can really be used as some kind of a metadata that describes what kind of a batch should be returned from the dataset. That's fair, and I don't want to block this PR any further, but we might still want to think about the naming, because it's not really an index.